### PR TITLE
Fix/add qrcode support

### PR DESCRIPTION
**Pull request by anestos**
Added support for QR code and fixed some issues with the unhappy flows (when user rejected or authentication expired etc)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -15,7 +15,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.curityVersion>4.4.0</project.curityVersion>
         <project.slf4jVersion>1.7.25</project.slf4jVersion>
-        <kotlin.version>1.2.50</kotlin.version>
+        <kotlin.version>1.3.72</kotlin.version>
```
Comment (travisspencer) [inline pom.xml]: The 5.0 docs say that the version that Curity provides is `1.3.61`. Are those out of date or is this version newer than it ought to be?

File: pom.xml [hunk 2]
```diff
@@ -71,6 +71,17 @@
...
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-test</artifactId>
+            <version>${kotlin.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-stdlib-jdk8</artifactId>
+            <version>${kotlin.version}</version>
```
Comment (travisspencer) [inline pom.xml]: is this a duplicate of the above? Only the scope is different. Should this have `test` scope?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        if (_minRegistrationLevel == RegistrationLevel.BASIC)
+        {
+            dataMap["attributesToReturn"] = _attributesToReturn.filter { attr ->
+                AttributesToReturn.EMAIL_ADDRESS == attr || AttributesToReturn.INTEGRATOR_SPECIFIC_USER_ID == attr
+            }.map { mapOf("attribute" to it) }
+        }
+        else if (_attributesToReturn.isNotEmpty() && _minRegistrationLevel != RegistrationLevel.BASIC)
+        {
+            dataMap["attributesToReturn"] = _attributesToReturn.map { mapOf("attribute" to it) }
+        }
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt:78]: no else case?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    fun getAuthTransaction(postData: Map<String, Any>): Map<String, Any>
+    {
+        val initialRequest = "initAuthRequest=${Base64.getEncoder().encodeToString(buildJsonAuthRequest(postData).toByteArray())}"
+        val requestBody = _relyingPartyId?.let { "$initialRequest&relyingPartyId=$it" } ?: initialRequest
+        val httpResponse = getWebServiceClient(config.environment.getHost())
+                .withPath("/authentication/1.0/initAuthentication")
+                .request()
+                .contentType("text/plain")
+                .body(HttpRequest.fromString(requestBody, StandardCharsets.UTF_8))
+                .method("POST")
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt]: can change to `.post()` IINM. Nitpick sorry. Not an important change.

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+        if (statusCode != 200)
+        {
+            if (_logger.isWarnEnabled)
+            {
+                _logger.warn("Got error response from authentication endpoint: error = {}, {}", statusCode,
+                        httpResponse.body(HttpResponse.asString()))
+            }
+
+            if (_json.fromJson(httpResponse.body(HttpResponse.asString()))["code"] == 2000)
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt:104]: 2000 isn't a typo? Shouldn't be 200?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+            throw _exceptionFactory.internalServerException(ErrorCode.EXTERNAL_SERVICE_ERROR)
+        }
+
+        return _json.fromJson(httpResponse.body(HttpResponse.asString()))
+    }
+
+    private fun buildJsonAuthRequest(postData: Map<String, Any>): String
+    {
+        return "{" + postData.map {
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt:117]: 👽

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 7]
```diff
@@ -84,13 +69,24 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
+
+            if (config.qrCodeEnabled())
             {
-                startAuthentication(requestModel)
+                //if qrcode is enabled, throw to go to wait handler
+                // and skip the form entry for email, ssn or phone
+                throw _exceptionFactory.redirectException(getWaitHandlerUrl(),
+                        RedirectStatusCode.MOVED_TEMPORARILY, emptyMap(), false)
             }
             else
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt]: else if ... else

why the nesting of the if else in the outer if?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 8]
```diff
@@ -131,82 +127,6 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
-                throw _exceptionFactory.badRequestException(ErrorCode.EXTERNAL_SERVICE_ERROR, "too.many.attempts")
-            }
-
-            throw _exceptionFactory.internalServerException(ErrorCode.EXTERNAL_SERVICE_ERROR)
-        }
-
-        return _json.fromJson(httpResponse.body(HttpResponse.asString()))
-    }
-
-    private fun buildJsonAuthRequest(postData: Map<String, Any>): String
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:166]: ah, that code I commented on above was a move. NM then.

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt [hunk 9]
```diff
@@ -61,21 +57,42 @@ class WaitRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig)
...
+            var viewData = emptyMap<String, String>()
+            if (config.qrCodeEnabled())
+            {
+                //the following data are proposed from freja documentation on QRCode generation
+                val postData = mutableMapOf("userInfoType" to "INFERRED", "userInfo" to "N/A")
+                val responseData = _requestLogicHelper.getAuthTransaction(postData)
+                val authRef = responseData["authRef"]?.toString()
+                config.sessionManager.put(Attribute.of(SESSION_AUTH_REF, authRef))
+
+                var cspOverride = "img-src 'self' " + if (config.environment == PredefinedEnvironment.PRODUCTION) QR_CODE_GENERATE_URL_PROD else QR_CODE_GENERATE_URL_TEST
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt]: we have this check in the definition of the URL, right?
https://github.com/curityio/freja-eid-authenticator/pull/11/files#diff-6e22681932fc980c5698748905d44dd7R122

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/config/FrejaEidAuthenticatorPluginConfig.kt [hunk 10]
```diff
@@ -17,7 +17,9 @@
 package io.curity.identityserver.plugin.frejaeid.config;
 
 import se.curity.identityserver.sdk.config.Configuration
+import se.curity.identityserver.sdk.config.annotation.DefaultBoolean
 import se.curity.identityserver.sdk.config.annotation.Description
+import se.curity.identityserver.sdk.config.annotation.Name
 import se.curity.identityserver.sdk.service.*
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/config/FrejaEidAuthenticatorPluginConfig.kt]: these `*` should be expanded

File: src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm [hunk 11]
```diff
@@ -0,0 +1,30 @@
+#*
+*  Copyright 2018 Curity AB
```
Comment (travisspencer) [inline src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm]: 2020

File: src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        #message("authenticator.freja-eid.authenticate.${error}")<br>
+    </div>
+</div>
+<div class="center py2 login-actions">
+    <a href="$_authUrl" role="button">
+        <i class="icon ion-refresh inlineicon"></i>#message("authenticator.freja-eid.authenticate.error.restart")</a>
+</div>
+#end
+
+#parse("layouts/default")
```
Comment (travisspencer) [inline src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm:30]: they error layout shouldn't be used

**Review by travisspencer — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (anestos) [reply pom.xml]: True, changing that

File: pom.xml (hunk 2 above)
Comment (anestos) [reply pom.xml]: removing

File: src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm (hunk 11 above)
Comment (anestos) [reply src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm]: fixed

File: src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm (hunk 12 above)
Comment (anestos) [reply src/main/resources/templates/authenticator/freja-eid/authenticate/error.vm:30]: It is a known error, when the user has declined the login in the app, or when the authentication has expired etc. I want the users to see the error and be able to restart, so this is the correct choice.

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/config/FrejaEidAuthenticatorPluginConfig.kt (hunk 10 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/config/FrejaEidAuthenticatorPluginConfig.kt]: fixed

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt (hunk 5 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt:104]: 2000 means "Authentication request failed. Previous authentication request was rejected due to security reasons."
see https://frejaeid.com/rest-api/Authentication%20Service.html

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt (hunk 4 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt]: Fixing in all places we do this

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt (hunk 3 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/RequestLogicHelper.kt:78]: Not needed, the `else` would not add any `attributesToReturn`

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt (hunk 9 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt]: fixed

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt (hunk 7 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt]: fixed

**Comment by anestos** [comment]
@travisspencer should the version be 2.2.0 or 3.0.0? (since any template changes are breaking)

**Comment by travisspencer** [comment]
Could be good to use a constant and add that link to the javadocs of that

**Comment by travisspencer** [comment]
Go for 3.0 since it's breaking

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt [hunk 13]
```diff
@@ -61,21 +57,43 @@ class WaitRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig)
...
-            response.setResponseModel(ResponseModel.templateResponseModel(emptyMap(), "authenticate/wait"),
+            var viewData = emptyMap<String, String>()
+            if (config.qrCodeEnabled())
+            {
+                //the following data are proposed from freja documentation on QRCode generation
+                val postData = mutableMapOf("userInfoType" to "INFERRED", "userInfo" to "N/A")
+                val responseData = _requestLogicHelper.getAuthTransaction(postData)
+                val authRef = responseData["authRef"]?.toString()
+                config.sessionManager.put(Attribute.of(SESSION_AUTH_REF, authRef))
+                val baseUrl = if (config.environment == PredefinedEnvironment.PRODUCTION) QR_CODE_GENERATE_URL_PROD else QR_CODE_GENERATE_URL_TEST
```
Comment (atifsaddique211f) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt:87]: why not a ternary operator rather than if/else?

**Review by atifsaddique211f — APPROVED** [review]

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt (hunk 13 above)
Comment (travisspencer) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt:87]: I think that code was removed.

**Pull request merged**